lib/rubygems/resolver.rb

```diff
@@ -277,6 +277,40 @@ def allow_missing?(dependency)
     @soft_missing
   end
 
+  def sort_dependencies(dependencies, activated, conflicts)
+      dependencies.sort_by do |dependency|
+        name = name_for(dependency)
+        [
+          activated.vertex_named(name).payload ? 0 : 1,
+          amount_constrained(dependency),
+          conflicts[name] ? 0 : 1,
+          activated.vertex_named(name).payload ? 0 : search_for(dependency).count,
+        ]
+      end
+    end
+
+  # returns an integer \in (-\infty, 0]
+  # a number closer to 0 means the dependency is less constraining
+  #
+  # dependencies w/ 0 or 1 possibilities (ignoring version requirements)
+  # are given very negative values, so they _always_ sort first,
+  # before dependencies that are unconstrained
+  def amount_constrained(dependency)
+    @amount_constrained ||= {}
+    @amount_constrained[dependency.name] ||= begin
+      dependency_request_for_name = Gem::Resolver::DependencyRequest.new(Gem::Dependency.new(dependency.name), dependency.requester)
+      all = @set.find_all(dependency_request_for_name).size
+
+      if all <= 1
+        all - 1_000_000
```

Comment:
it's a sufficiently large "boost" for dependencies where the sum of all possibilities (even disregarding any requirements) is either 0 or 1, which I tried to explain in the comment above the method